tests/state/state.go

```diff
@@ -0,0 +1,301 @@
+	// branchable collections
+	ActualDAGHeads map[string]DocHeadState
+
+	// ExpectedDAGHeads contains all DAG heads that are expected to exist on a node.
+	//
+	// The map key is the doc id. The map value is the DAG head.
+	//
+	// This tracks composite commits for documents, and collection commits for
+	// branchable collections
+	ExpectedDAGHeads map[string]cid.Cid
+}
+
+// DocHeadState contains the state of a document head.
+// It is used to track if a document at a certain head has been decrypted.
+type DocHeadState struct {
+	// The actual document head.
+	Cid cid.Cid
+	// Indicates if the document at the given head has been Decrypted.
+	Decrypted bool
+}
+
+// NewP2PState returns a new empty p2p state.
+func NewP2PState() *P2pState {
+	return &P2pState{
+		Connections:      make(map[int]struct{}),
+		Replicators:      make(map[int]struct{}),
+		PeerCollections:  make(map[int]struct{}),
+		PeerDocuments:    make(map[ColDocIndex]struct{}),
+		ActualDAGHeads:   make(map[string]DocHeadState),
+		ExpectedDAGHeads: make(map[string]cid.Cid),
+	}
+}
+
+// EventState contains all event related testing state for a node.
+type EventState struct {
+	// Merge is the `event.MergeCompleteName` subscription
+	Merge event.Subscription
+
+	// Update is the `event.UpdateName` subscription
+	Update event.Subscription
+
+	// Replicator is the `event.ReplicatorCompletedName` subscription
+	Replicator event.Subscription
+}
+
+// NewEventState returns an eventState with all required subscriptions.
+func NewEventState(bus event.Bus) (*EventState, error) {
+	merge, err := bus.Subscribe(event.MergeCompleteName)
+	if err != nil {
+		return nil, err
+	}
+	update, err := bus.Subscribe(event.UpdateName)
+	if err != nil {
+		return nil, err
+	}
+	replicator, err := bus.Subscribe(event.ReplicatorCompletedName)
+	if err != nil {
+		return nil, err
+	}
+	return &EventState{
+		Merge:      merge,
+		Update:     update,
+		Replicator: replicator,
+	}, nil
+}
+
+// NodeState contains all testing state for a node.
+type NodeState struct {
+	// The node's client active in this test.
+	clients.Client
+	// Event contains all Event node subscriptions.
+	Event *EventState
+	// P2p contains P2p states for the node.
+	P2p *P2pState
+	// The network configurations for the nodes
+	NetOpts []netConfig.NodeOpt
+	// The path to any file-based databases active in this test.
+	DbPath string
+	// Collections by index present in the test.
+	// Indexes matches that of collectionNames.
+	Collections []client.Collection
+	// indicates if the node is Closed.
+	Closed bool
+	// AddrInfo contains the peer information for the node.
+	AddrInfo peer.AddrInfo
+}
+
+// State contains all testing State.
+type State struct {
+	// The test context.
+	Ctx context.Context
+
+	// The Go Test test state
+	T testing.TB
+
+	// The type of KMS currently being tested.
+	Kms KMSType
+
+	// The type of database currently being tested.
+	Dbt DatabaseType
```

Comment:
suggestion: This one is a bit cryptic in this form. Maybe `DBType` would clearer for readers.

Reply:
I've renamed this as suggested